core/src/main/java/io/ultrabrew/metrics/data/BasicHistogramAggregator.java

```diff
@@ -0,0 +1,266 @@
+  }
+
+  /**
+   * Creates a monoid for the histogram buckets for a metric.
+   *
+   * @param metricId identifier of the metric associated with this aggregator
+   * @param buckets distribution bucket spec
+   * @param maxCardinality requested max capacity of table in records. Table doesn't grow beyond
+   * @param cardinality requested capacity of table in records, actual capacity may be higher
+   */
+  public BasicHistogramAggregator(final String metricId, final DistributionBucket buckets,
+      final int maxCardinality, final int cardinality) {
+    super(AGG_FIELDS.length + buckets.getCount(), maxCardinality, cardinality, IDENTITY);
+    this.metricId = metricId;
+    this.buckets = buckets;
+    this.bucketCount = buckets.getCount();
+    this.fields = buildFields();
+    this.types = buildTypes(fields.length);
+    this.identity = buildIdentity(fields.length);
+  }
+
+  @Override
+  public void apply(final String[] tags, final long value, final long timestamp) {
+    super.apply(tags, value, timestamp);
+  }
+
+  @Override
+  protected void combine(final int[] table, final long baseOffset, final long value) {
+    addToAggField(table, baseOffset, 0, 1);
+    addToAggField(table, baseOffset, 1, value);
+    minAggField(table, baseOffset, 2, value);
+    maxAggField(table, baseOffset, 3, value);
+    setAggField(table, baseOffset, 4, value);
+
+    //Increments the bucket counter by 1 responding to the given value
+    int bucketIndex = buckets.getBucketIndex(value);
+    addToDataField(table, baseOffset, AGG_FIELDS.length + bucketIndex, 1);
+  }
+
+  @Override
+  public Cursor cursor() {
+    return newCursor(false);
+  }
+
+  @Override
+  public Cursor sortedCursor() {
+    return newCursor(true);
+  }
+
+  private Cursor newCursor(boolean sorted) {
+    return new CursorImpl(tagSets, fields, types, identity, sorted);
+  }
+
+  /**
+   * Creates an array of identity values of the aggregation field and histogram buckets. These
+   * values are used to initialize and reset the field after reading.
+   */
+  private long[] buildIdentity(int length) {
+    long[] identity = new long[length];
+    System.arraycopy(IDENTITY, 0, identity, 0, IDENTITY.length);
+    return identity;
+  }
+
+  /**
+   * Creates an array of names of the aggregation field and histogram buckets.
+   */
+  private String[] buildFields() {
+    String[] fields = new String[AGG_FIELDS.length + bucketCount];
+    String[] buckets = this.buckets.getBucketNames();
+    System.arraycopy(AGG_FIELDS, 0, fields, 0, AGG_FIELDS.length);
+    System.arraycopy(buckets, 0, fields, AGG_FIELDS.length, buckets.length);
+    return fields;
+  }
+
+  private Type[] buildTypes(final int length) {
+    Type[] types = new Type[length];
+    Arrays.fill(types, Type.LONG);
+    return types;
+  }
+
+  private class CursorImpl implements Cursor {
+
+    private final String[] fields;
+    private final Type[] types;
+    private final long[] identity;
```

Comment:
Since this is a non-static inner class, `fields`, `types` and `identity` can be accessed directly from the outer class, no need to have duplicate variables here.